### PR TITLE
Script for data import and export (e.g. pushing to an "open science" LORIS instance)

### PR DESCRIPTION
**Pull request by johnsaigle**
## Brief summary of changes

This script was originally written for the use of getting a subset of PreventAD data from its instance into a new "Open Science" instance. It has been thoroughly reviewed and tested by @Jkat and @cmadjar. 

I'm pushing this to the main repo as @samirdas mentioned that it should go here. It's probable that other projects in the future will need something similar and may benefit from this script.

Further background and detail can be found in the discussion on the original PR on my LORIS fork: https://github.com/johnsaigle/Loris/pull/15

The original PR description follows.

---

This pair of scripts allows one to export data from one database and import it another database, mapping PSCIDs from the former to new values. This allows for a way of transferring data into e.g. an Open Science database while giving the candidates new identifier to preserve privacy between databases.

## Data Extraction
The dataExtractor.php script is used to extract data from the source database. It can extract data from either the candidate table, the session table, or from tables containing instrument data. Other tables are unsupported and the behaviour of this script when trying to extract data from other tables is undefined/almost guaranteed to have bugs.

The script writes data to project/data_export/ in CSV format. Optionally can be written elsewhere.

## Data Import
The openScienceDataImport.php script is used to import data. It takes several parameters:

An operation mode, either 'candidate', 'visits', or 'instrument', corresponding to the dataExtractor tool
A CSV file mapping old/source PSCIDs to a new PSCID and new CandID for the destination database.
A source data file generated by dataExtractor.php. This file should not be renamed as the name of the MySQL table to update will be included in the name
It analyzes the two CSV files and generates SQL commands, either INSERT or UPDATE, that can be used to insert the data into the database. These commands are written to a file in project/data_import/ with a timestamp.

Candidate information will always be UPDATE statements as the candidates are expected to be created in the candidate table beforehand.
Visits information will be a mix of UPDATE or INSERT depending on whether the visit label exists
Instrument information will always be INSERTs as it is expected that instrument information is not yet populated in the DB.

### CODE REVIEW

**Comment by johnsaigle** [comment]
TODO

- [ ] Must pass Travis
- [ ] Delete sample CSV files
- [ ] Better document its usage (maybe a wiki page)

File: tools/dataExtractor.php [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            'Please use the `visits` extraction mode to retrive information ' .
+            'from the session table.' . PHP_EOL
+        );
+    }
+
+    // A single column or list of columns joined by commas, e.g. "DoB,Sex"
+    $column = $argv[3];
+
+    $cutoffDate = $argv[4];
+    //
```
Comment (cmadjar) [inline tools/dataExtractor.php]: remove empty comment?

File: tools/dataExtractor.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        // Create an array containing 'PSCID' and the column names from the
+        // command line input. Then join them as a comma-separated string (to be
+        // written as the headers to the CSV file).
+        $headers = array_merge(array('PSCID'), explode(',', $column));
+
+        // Build basic SQL query info.
+        $query = "SELECT PSCID,$column from $table";
+        $params = array();
+    } else if ($mode === INSTRUMENT_EXPORT) {
+        // Query whether a tabel has the `Date taken` column.
```
Comment (cmadjar) [inline tools/dataExtractor.php]: typo in table

File: tools/dataExtractor.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
+<?php declare(strict_types=1);
```
Comment (cmadjar) [inline tools/dataExtractor.php]: shouldn't this script be moved to the `exporters` folder within the `tools` directory?

File: tools/dataExtractor.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+/**
+ * The LORIS subfolder where output file will be written.
+ *
+ * @var string
+ */
+const OUTPUT_FOLDER = 'project/data_export/';
+
+
+// TODO Make Usage message more informative.
```
Comment (cmadjar) [inline tools/dataExtractor.php]: hehe, would be nice to do it ;)

File: tools/openScienceDataImporter.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+const NUM_ARGS_REQUIRED = 4;
+
+/*
+ * Expected indices of parameters passed to this script via $argv
+ *
+ * @var int
+ */
+const MAPPING_ARG_INDEX = 2;
+const DATA_ARG_INDEX = 3;
+const EXCLUDED_ARG_INDEX = 4; // Used only for visit label imports
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: phpcs?

File: tools/openScienceDataImporter.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+/**
+ * Command line argument for the mode of data importation.
+ */
+const CANDIDATE_IMPORT = 'candidate';
+const INSTRUMENT_IMPORT = 'instrument';
+const VISIT_IMPORT = 'visits';
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: phpcs?

File: tools/openScienceDataImporter.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    <data.csv>      A CSV file containing columns with OLDPSCID,DATA
+
+
+To import instrument information specifically:
+
+php {$argv[0]} %s <mapping.csv> <data.csv>
+    <mapping.csv>   A CSV file containg columns with OLDPSCID, NEWPSCID, and
+                    NEWCANDID
+    <data.csv>      A CSV file containing columns with instrument data. This
+                    file should be created using the dataExtractor tool. The
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: if the `dataExtractor.php` tool is moved to the `exporters` folder, will need to update the description here accordingly

File: tools/openScienceDataImporter.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                    filename must begin with the name of the instrument table
+                    followed by an underscore.
+
+
+To import session information specifically:
+
+php {$argv[0]} %s <mapping.csv> <visits.csv> [excluded.csv]
+    <mapping.csv>   A CSV file containg columns with OLDPSCID, NEWPSCID, and
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: typo in `containing`

File: tools/openScienceDataImporter.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+php {$argv[0]} %s <mapping.csv> <data.csv>
+    <mapping.csv>   A CSV file containg columns with OLDPSCID, NEWPSCID, and
+                    NEWCANDID
+    <data.csv>      A CSV file containing columns with OLDPSCID,DATA
+
+
+To import instrument information specifically:
+
+php {$argv[0]} %s <mapping.csv> <data.csv>
+    <mapping.csv>   A CSV file containg columns with OLDPSCID, NEWPSCID, and
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: typo in `containing`

File: tools/openScienceDataImporter.php [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        implode("\n", array_map('formatBulletPoint', $importer->candidatesSkipped)) .
+        PHP_EOL . PHP_EOL;
+}
+
+// Write the report to file.
+$path = $config->getSetting('base') . 'project/data_import/'; 
+if (!is_dir($path)) {
+    mkdir($path);
+}
+$target = $path . date("Ymd-His") . '_dataImporterOutput.sql';
```
Comment (cmadjar) [inline tools/openScienceDataImporter.php]: should the name of the table or instrument be in the SQL file produced?

File: tools/CandidateImporter.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
+<?php declare(strict_types=1);
```
Comment (cmadjar) [inline tools/CandidateImporter.php]: should there be a folder created in the tools directory for the importer scripts? Same as the `exporters` directory already there.

Comment (cmadjar) [reply tools/CandidateImporter.php]: @ridz1208

File: tools/DataImporter.class.inc [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+     */
+    const OLD_PSCID = 0;
+    const NEW_PSCID = 1;
+    const NEW_CANDID = 2;
+
+    /* An associative array of CSV headers to rows of CSV data */
+    protected $dataRows = array();
+
+    /* An associative array of CSV headers to rows of CSV data. Contains a
+     * mapping of the original PSCIDs used in the source databse to the
```
Comment (cmadjar) [inline tools/DataImporter.class.inc]: typo in `database`

File: tools/DataImporter.class.inc [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+    private $candIDMapping = array();
+
+    /* An array of PSCIDs present in the candidate table of the database. */
+    private $existingPSCIDs = array();
+
+    /* An array of PSCIDs that are present in the mapping file but skipped for
+     * processing. */
+    public $candidatesSkipped = array();
+
+    /* An array containing informaiton to be formatted into a MySQL UPDATE
```
Comment (cmadjar) [inline tools/DataImporter.class.inc]: typo in information

File: tools/DataImporter.class.inc [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+     * statement. */
+    protected $UPDATEQueue = array();
+
+    /* An array containing informaiton to be formatted into a MySQL INSERT
```
Comment (cmadjar) [inline tools/DataImporter.class.inc]: typo in information

File: tools/VisitImporter.php [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                $this->excludedCount++;
+                return;
+            }
+        }
+
+        $data = $row;
+        
+        
+        // Prepare command information.
+        //
```
Comment (cmadjar) [inline tools/VisitImporter.php]: empty comment

**Comment by johnsaigle** [comment]
> looks like all the scripts need to be run through phpcbf (probably failing Travis I guess)

Tools don't go through PHPCS yet actually, but they will after #5766. Anyway, this is passing 😄 

> missing documentation on all scripts. Would be good to have them done

Agreed, I'll keep the Needs Documentation label on this PR until I have time to write and commit documentation.

Thanks for reviewing. I'll fix the typos and bug soon.

File: tools/CandidateImporter.php (hunk 11 above)
Comment (ridz1208) [reply tools/CandidateImporter.php]: same, we can rename it importers/exporters

**Comment by johnsaigle** [comment]
@cmadjar Could you re-review this PR? I think I've incorporated all of your suggestions. It's now also passing PHPCS.

File: tools/importers/DataImporter.class.inc [hunk 16]
```diff
@@ -0,0 +1,489 @@
...
+    {
+        return $this->candIDMapping[$newPSCID] ?? null;
+    }
+
+
+    /**
+     * Returns a visual representation of the SQL queries to be run in order to
+     * perform data import. Used for displaying to users and logging.
+     *
+     * @return string[] SQL commands to be run to perform import.
```
Comment (johnsaigle) [inline tools/importers/DataImporter.class.inc:247]:
```suggestion
     * @return string SQL commands to be run to perform import (joined by new lines).
```

File: tools/importers/VisitImporter.php [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        $this->table = 'session';
+
+        // Get existing sessions so that there is a way to distinguish between
+        // when to build UPDATE vs. INSERT statements.
+        $result = \Database::singleton()->pselect(
+            'SELECT CandID,Visit_label FROM session',
+            array()
+        );
+        foreach ($result as $row) {
+            $this->existingSessions[$row['CandID']] = $row['Visit_label'];
```
Comment (cmadjar) [inline tools/importers/VisitImporter.php:40]:
```suggestion
            if (!array_key_exists($row['CandID'], $this->existingSessions)) {
                $this->existingSessions[$row['CandID']] = array();
            }
            array_push(
                $this->existingSessions[$row['CandID']],
                $row['Visit_label']
            );
```

File: tools/importers/VisitImporter.php [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        // We don't want PSCID information from the CSV file included in the
+        // SET statement. It's only used for linking.
+        unset($data['PSCID']);
+        $where   = array('CandID' => $newCandID);
+        $command = array(
+            'data'  => $data,
+            'where' => $where,
+        );
+
+        if (!empty($this->existingSessions[$newCandID])
```
Comment (cmadjar) [inline tools/importers/VisitImporter.php:105]:
```suggestion
        if (empty($this->existingSessions[$newCandID])
```

**Pull request merged**